src/servant/tt_002/3678.c

```diff
@@ -1029,7 +1034,210 @@ void func_us_80176B6C(Entity* self) {
     ServantUpdateAnim(self, 0, D_us_80172B14);
 }
 
-INCLUDE_ASM("servant/tt_002/nonmatchings/3678", func_us_80176C98);
+void func_us_80176C98(Entity* self) {
+    s32 rnd;
+    s32 i;
+
+    g_api.GetServantStats(self, 0, 0, &s_FaerieStats);
+    if (D_us_80179320 != 0) {
+        self->zPriority = PLAYER.zPriority - 2;
+        s_zPriority = self->zPriority;
+    }
+    if (IsMovementAllowed(1)) {
+        if (self->step < 2) {
+            self->entityId = 0xD1;
+            self->step = 0;
+            return;
+        }
+
+        if (self->step < 9) {
+            self->step = 9;
+        }
+    }
+    if (PLAYER.step_s == 0) {
+        s_TargetLocOffset_calc = -6;
+        if (PLAYER.facingLeft) {
+            s_TargetLocOffset_calc = -s_TargetLocOffset_calc;
+        }
+        s_TargetLocationX = PLAYER.posX.i.hi + s_TargetLocOffset_calc;
+        s_TargetLocationY = PLAYER.posY.i.hi - 12;
+    } else {
+        s_TargetLocOffset_calc = 16;
+        if (PLAYER.facingLeft) {
+            s_TargetLocOffset_calc = -s_TargetLocOffset_calc;
+        }
+        s_TargetLocationX = PLAYER.posX.i.hi + s_TargetLocOffset_calc;
+        s_TargetLocationY = PLAYER.posY.i.hi - 8;
+    }
+    switch (self->step) {
+    case 0:
+        func_us_801739D0(self);
+        func_us_80173994(self, 0xE);
+        break;
+    case 1:
+        s_AngleToTarget = CalculateAngleToEntity(self, s_TargetLocationX, s_TargetLocationY);
+        s_AllowedAngle = GetTargetPositionWithDistanceBuffer(s_AngleToTarget, self->ext.faerie.targetAngle, 0x180);
+        self->ext.faerie.targetAngle = s_AllowedAngle;
+        self->velocityY = -(rsin(s_AllowedAngle) << 3);
+        self->velocityX = (rcos(s_AllowedAngle) << 3);
+        func_us_80173BD0(self);
+        self->posX.val += self->velocityX;
+        self->posY.val += self->velocityY;
+        s_DistToTargetLocation = CalculateDistance(self, s_TargetLocationX, s_TargetLocationY);
+        if (s_DistToTargetLocation < 2) {
+            if (PLAYER.step_s == 0) {
+                self->facingLeft = PLAYER.facingLeft;
+            } else {
+                self->facingLeft = PLAYER.facingLeft ? 0 : 1;
+            }
+            func_us_80173994(self, 0x18);
+            self->ext.faerie.frameCounter = 0;
+            
+            self->flags |= FLAG_POS_PLAYER_LOCKED;
+            self->flags &= ~FLAG_POS_CAMERA_LOCKED;
+
+#ifdef VERSION_PSP
+            if (D_8003C708.flags & LAYOUT_RECT_PARAMS_UNKNOWN_20
+                || D_8003C708.flags & LAYOUT_RECT_PARAMS_UNKNOWN_40 
+                || (D_us_8017931C == 1) 
+                || s_FaerieStats.level < 5) {
+                self->step = 5;
+            } else if (s_ServantId != FAM_STATS_NOSE_DEMON || PLAYER.step_s != 4){
```

Comment:
use the FamiliarActiveIds for comparing against s_ServantId.  IE FAM_STATS_NOSE_DEMON should be FAM_ACTIVE_YOUSEI